File: .net/HarvbotArms/Harvbot.Arms.Driver/HarvbotArmNode.cs
using System;
using System.Linq;

namespace Harvbot.Arms.Driver
{
    /// <summary>
    /// Represents Harvbot Arm node.
    /// </summary>
    public class HarvbotArmNode
    {
        /// <summary>
        /// Represents Harvbot Arm node.
        /// </summary>
        /// <param name="type">The node type.</param>
        /// <param name="arm">The owner.</param>
        public HarvbotArmNode(HarvbotArmNodeTypes type, HarvbotArmBase arm)
        {
            this.Arm = arm;
            this.Type = type;
        }

        /// <summary>
        /// Gets the node owner.
        /// </summary>
        public HarvbotArmBase Arm { get; private set; }

        /// <summary>
        /// Gets the node type.
        /// </summary>
        public HarvbotArmNodeTypes Type { get; private set; }

        /// <summary>
        /// Moves node to specified position.
        /// </summary>
        /// <param name="degree">The degree.</param>
        public void Move(int degree)
        {
            var response = this.SendCommand("move", degree.ToString());

            if (string.IsNullOrEmpty(response))
            {
                throw new InvalidOperationException($"Invalid moving: {degree}");
            }
        }

        /// <summary>
        /// Sweep node to specified position.
        /// </summary>
        /// <param name="degree">The degree.</param>
        public void Sweep(int degree)
        {
            var response = this.SendCommand("sweep", degree.ToString());

            if (string.IsNullOrEmpty(response))
            {
                throw new InvalidOperationException($"Invalid moving: {degree}");
            }
        }

        /// <summary>
        /// Gets current node position.
        /// </summary>
        /// <returns>The current node position degree.</returns>
        public int? GetPosition()
        {
            var response = this.SendCommand("pos");

            if (!string.IsNullOrEmpty(response))
            {
                return int.Parse(response);
            }

            return new int?();
        }

        /// <summary>
        /// Sends commands to controller.
        /// </summary>
        /// <param name="command">The command.</param>
        /// <param name="args">The command arguments.</param>
        /// <returns>The response.</returns>
        protected string SendCommand(string command, string args = null)
        {
            if (string.IsNullOrEmpty(args))
            {
                this.Arm.SerialPort.WriteLine($"harm:{command}:{(int)this.Type}:~harm");
            }
            else
            {
                this.Arm.SerialPort.WriteLine($"harm:{command}:{(int)this.Type}:{args}:~harm");
            }

            var response = this.Arm.SerialPort.ReadLine();

            var segments = response.Split(new string[] { ":" }, StringSplitOptions.RemoveEmptyEntries);

            if (segments[0].Equals("harm", StringComparison.InvariantCultureIgnoreCase) &&
                segments[1].Equals(command, StringComparison.InvariantCultureIgnoreCase) &&
                segments[2].Equals(((int)this.Type).ToString(), StringComparison.InvariantCultureIgnoreCase) &&
                segments.Last().Equals("~harm", StringComparison.InvariantCultureIgnoreCase))
            {
                throw new InvalidOperationException($"Invalid response: {response}");
            }

            return segments[3];
        }
    }
}

File: android/arms/src/main/java/com/harvbot/arms/HarvbotArmNode.java
package com.harvbot.arms;

import android.text.TextUtils;

import com.hoho.android.usbserial.driver.UsbSerialPort;

import java.io.IOException;
import java.io.UnsupportedEncodingException;

public class HarvbotArmNode
{
    private HarvbotArmBase arm;

    private HarvbotArmNodeTypes type;

    public static final int Timeout = 1000;

    public HarvbotArmNode(HarvbotArmNodeTypes type, HarvbotArmBase arm)
    {
        this.arm = arm;
        this.type = type;
    }

    public HarvbotArmNodeTypes getType()
    {
        return this.type;
    }

    public void move(int degree) throws IOException
    {
        this.sendCommand("move", String.valueOf(degree));
    }

    public void sweep(int degree) throws IOException
    {
        this.sendCommand("sweep", String.valueOf(degree));
    }

    public Integer getPosition() throws IOException
    {
        String response = this.sendCommand("pos");

        if (!TextUtils.isEmpty(response))
        {
            return Integer.parseInt(response);
        }

        return null;
    }

    protected String sendCommand(String command) throws IOException
    {
        return this.sendCommand(command, null);
    }

    protected String sendCommand(String command, String args) throws IOException
    {
        UsbSerialPort serailPost = this.arm.getSerialPort();
        StringBuilder request = new StringBuilder();

        request.append("harm:" + command + ":");
        request.append(String.valueOf(this.type.getNumericType()));
        if (!TextUtils.isEmpty(args))
        {
            request.append(":" + args);
        }

        request.append(":~harm\n");

        byte[] requestBytes = request.toString().getBytes("ASCII");
        serailPost.write(requestBytes, Timeout);

        byte[] responseBytes = new byte[128];
        int read = serailPost.read(responseBytes, Timeout);

        String response = new String(responseBytes, "ASCII");

        response = response.trim();

        if (!TextUtils.isEmpty(response))
        {
            String[] segments = response.split(":");

            if (segments[0].equalsIgnoreCase("harm") &&
                    segments[1].equalsIgnoreCase(command) &&
                    segments[2].equalsIgnoreCase(String.valueOf((int) this.type.getNumericType())) &&
                    segments[segments.length - 1].equalsIgnoreCase("~harm")) {
                return segments[3];
            }
        }

        return "";
    }
}